psl-core/src/main/java/org/linqs/psl/evaluation/statistics/RankingEvaluator.java

```diff
@@ -21,255 +21,254 @@
 
     public RankingEvaluator() {
-        this(Options.EVAL_RANKING_THRESHOLD.getDouble());
+        this(RepresentativeMetric.MRR);
     }
 
-    public RankingEvaluator(double threshold) {
-        this(threshold, Options.EVAL_RANKING_REPRESENTATIVE.getString());
+    public RankingEvaluator(RepresentativeMetric representative) {
+        this.representative = representative;
+        this.categoryIndex = Options.EVAL_RANK_CATEGORY_INDEX.getInt();
+        this.threshold = Options.EVAL_RANK_THRESHOLD.getDouble();
+        this.categoryMap = new HashMap<Constant, ArrayList<GroundAtom>>();
+        defaultPredicate = Options.EVAL_RANK_DEFAULT_PREDICATE.getString();
     }
 
-    public RankingEvaluator(double threshold, String representative) {
-        this(threshold, RepresentativeMetric.valueOf(representative.toUpperCase()));
-    }
+    /*
+    Grab the Constant argument of an atom in a specified position.
+    */
 
-    public RankingEvaluator(double threshold, RepresentativeMetric representative) {
-        if (threshold < 0.0 || threshold > 1.0) {
-            throw new IllegalArgumentException("Threhsold must be in (0, 1). Found: " + threshold);
-        }
+    public Constant getArgAtPosition(GroundAtom atom, int position) {
+        Constant[] args = atom.getArguments();
+        return args[position];
+    }
 
-        this.threshold = threshold;
-        this.representative = representative;
+    /*
+    Convert the Iterables returned by TrainingMap methods
+    into Lists.
+    */
 
-        truth = new ArrayList<GroundAtom>();
-        predicted = new ArrayList<GroundAtom>();
+    public static <E> ArrayList<E> iterableToList(Iterable<E> iterable) {
+        ArrayList<E> list = new ArrayList<E>();
+        for (E element : iterable) {
+            list.add(element);
+        }
+        return list;
     }
 
     @Override
     public void compute(TrainingMap trainingMap) {
-        compute(trainingMap, null);
+        if (defaultPredicate == null) {
+            throw new UnsupportedOperationException("RankingEvaluators must have a default predicate set (through config).");
+        }
+
+        compute(trainingMap, StandardPredicate.get(defaultPredicate));
     }
 
     @Override
     public void compute(TrainingMap trainingMap, StandardPredicate predicate) {
-        truth = new ArrayList<GroundAtom>(trainingMap.getLabelMap().size());
-        predicted = new ArrayList<GroundAtom>(trainingMap.getLabelMap().size());
-
-        for (Map.Entry<GroundAtom, GroundAtom> entry : getMap(trainingMap)) {
-            if (predicate != null && entry.getKey().getPredicate() != predicate) {
-                continue;
-            }
+        /*
+        Populate lists of all predicted/truth atoms
+        which have the specified predicate.
+        */
 
-            truth.add(entry.getValue());
-            predicted.add(entry.getKey());
-        }
+        predictedAtoms = iterableToList(trainingMap.getAllPredictions());
```

Comment:
I'm not convinced that turning the iterable into a list is a good thing here.
You are potentially using a relatively large amount of memory and two additional iterations to make two future iterations shorter.